packages/flutter/lib/src/cupertino/route.dart

```diff
@@ -1388,14 +1372,56 @@ class CupertinoDialogRoute<T> extends RawDialogRoute<T> {
     String? barrierLabel,
     // This transition duration was eyeballed comparing with iOS
     super.transitionDuration = const Duration(milliseconds: 250),
-    super.transitionBuilder = _buildCupertinoDialogTransitions,
+    this.transitionBuilder,
     super.settings,
     super.anchorPoint,
   }) : super(
         pageBuilder: (BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation) {
           return builder(context);
         },
+        transitionBuilder: transitionBuilder ?? _buildCupertinoDialogTransitions,
         barrierLabel: barrierLabel ?? CupertinoLocalizations.of(context).modalBarrierDismissLabel,
         barrierColor: barrierColor ?? CupertinoDynamicColor.resolve(kCupertinoModalBarrierColor, context),
       );
+
+  /// Custom transition builder
+  RouteTransitionsBuilder? transitionBuilder;
+
+  CurvedAnimation? _fadeAnimation;
+
+  @override
+  Widget buildTransitions(BuildContext context, Animation<double> animation, Animation<double> secondaryAnimation, Widget child) {
+
+    if (transitionBuilder != null) {
+      return super.buildTransitions(context, animation, secondaryAnimation, child);
+    }
+
+    if (_fadeAnimation?.parent != animation) {
+      _fadeAnimation?.dispose();
+      _fadeAnimation = CurvedAnimation(
+        parent: animation,
+        curve: Curves.easeInOut,
+      );
+    }
+
```

Comment:
Can you assert here that _fadeAnimation != null.
It will improve code readability and verify our assumptions are correct even after future changes.

Reply:
Actually it may be better to create local:

`final fadeAnimation = _fadeAnimation!`

And then use it without `!`

Reply:
Would like me to only add local variable change

> `final fadeAnimation = _fadeAnimation!`
> 
> And then use it without `!`

Or both assert and local variable change 

It's a bit confusing because of the below line:

> Actually it may be better to create a local:
>

Reply:
With local variable assert is not needed, because the expression contains '!', that will fail in case of null.
So, just local.